pkg/trace/traceutil/otel_util.go

```diff
@@ -0,0 +1,277 @@
+			if spanKind == ptrace.SpanKindServer || spanKind == ptrace.SpanKindConsumer {
+				// span is a server-side span, mark as top level
+				topLevelSpans[spanID] = struct{}{}
+			}
+		} else {
+			// Fall back to old behavior in ComputeTopLevel
+			if _, ok := spanByID[span.ParentSpanID()]; !ok {
+				// case 2: parent span not in the same chunk, presumably it belongs to another service
+				topLevelSpans[spanID] = struct{}{}
+			} else {
+				svc := GetOTelAttrVal(resByID[spanID].Attributes(), true, semconv.AttributeServiceName)
+				parentSvc := GetOTelAttrVal(resByID[span.ParentSpanID()].Attributes(), true, semconv.AttributeServiceName)
+				if svc != parentSvc {
+					// case 3: parent is not in the same service
+					topLevelSpans[spanID] = struct{}{}
+				}
+			}
+		}
+	}
+	return topLevelSpans
+}
+
+// GetOTelAttrVal returns the matched value as a string in the input map with the given keys.
+// If there are multiple keys present, the first matched one is returned.
+// If normalize is true, normalize the return value with NormalizeTagValue.
+func GetOTelAttrVal(attrs pcommon.Map, normalize bool, keys ...string) string {
+	val := ""
+	for _, key := range keys {
+		attrval, exists := attrs.Get(key)
+		if exists {
+			val = attrval.AsString()
+		}
+	}
+
+	if normalize {
+		val = NormalizeTagValue(val)
+	}
+
+	return val
+}
+
+// GetOTelAttrValInResAndSpanAttrs returns the matched value as a string in the OTel resource attributes and span attributes with the given keys.
+// If there are multiple keys present, the first matched one is returned.
+// If the key is present in both resource attributes and span attributes, resource attributes take precedence.
+// If normalize is true, normalize the return value with NormalizeTagValue.
+func GetOTelAttrValInResAndSpanAttrs(span ptrace.Span, res pcommon.Resource, normalize bool, keys ...string) string {
+	if val := GetOTelAttrVal(res.Attributes(), normalize, keys...); val != "" {
+		return val
+	}
+	return GetOTelAttrVal(span.Attributes(), normalize, keys...)
+}
+
+// GetOTelSpanType returns the DD span type based on OTel span kind and attributes.
+func GetOTelSpanType(span ptrace.Span, res pcommon.Resource) string {
+	var typ string
+	switch span.Kind() {
+	case ptrace.SpanKindServer:
+		typ = "web"
+	case ptrace.SpanKindClient:
+		db := GetOTelAttrValInResAndSpanAttrs(span, res, true, semconv.AttributeDBSystem)
+		if db == "redis" || db == "memcached" {
+			typ = "cache"
+		} else if db != "" {
+			typ = "db"
+		} else {
+			typ = "http"
+		}
+	default:
+		typ = "custom"
+	}
+	return typ
+}
+
+// GetOTelService returns the DD service name based on OTel span and resource attributes.
+func GetOTelService(span ptrace.Span, res pcommon.Resource, normalize bool) string {
+	// No need to normalize with NormalizeTagValue since we will do NormalizeService later
+	svc := GetOTelAttrValInResAndSpanAttrs(span, res, false, semconv.AttributeServiceName)
+	if svc == "" {
+		svc = "otlpresourcenoservicename"
```

Comment:
We were previously setting this to "OTLPResourceNoServiceName", is there any difference without the casing here?

Reply:
Good question, `traceutil.NormalizeService` always convert the service names to lower-case, so in the end `OTLPResourceNoServiceName` always ends up being `otlpresourcenoservicename`